src/WebJobs.Script/Description/ScriptFunctionInvokerBase.cs

```diff
@@ -64,27 +67,77 @@ protected virtual void OnScriptFileChanged(object sender, FileSystemEventArgs e)
+            ApplyAmbientBindingData(binder, bindingData);
+
+            // If there are any parameters in the bindings,
+            // get the binding data. In dynamic script cases we need
+            // to parse this POCO data ourselves - it won't be in the existing
+            // binding data because all the POCO binders require strong
+            // typing
+            if (outputBindings.Any(p => p.HasBindingParameters) ||
+                inputBindings.Any(p => p.HasBindingParameters))
             {
-                // parse the object skipping any nested objects (binding data
-                // only includes top level properties)
-                JObject parsed = JObject.Parse(value as string);
-                bindingData = parsed.Children<JProperty>()
-                    .Where(p => p.Value.Type != JTokenType.Object)
-                    .ToDictionary(p => p.Name, p => (string)p);
+                try
+                {
+                    string json = value as string;
+                    if (!string.IsNullOrEmpty(json) && IsJson(json))
+                    {
+                        // parse the object skipping any nested objects (binding data
+                        // only includes top level properties)
+                        JObject parsed = JObject.Parse(json);
+                        bindingData = parsed.Children<JProperty>()
+                            .Where(p => p.Value.Type != JTokenType.Object)
+                            .ToDictionary(p => p.Name, p => (string)p);
+                    }
+                }
+                catch
+                {
+                    // it's not an error if the incoming message isn't JSON
+                    // there are cases where there will be output binding parameters
+                    // that don't bind to JSON properties
+                }
             }
-            catch
+
+            return bindingData;
+        }
+
+        /// <summary>
+        /// TEMP HACK
+        /// We need to merge the ambient binding data that already exists in the IBinder
+        /// with our binding data. We have to do this rather than relying solely on
+        /// IBinder.BindAsync because we need to include any POCO values we get from parsing
+        /// JSON bodies, etc.
+        /// </summary>
+        protected static void ApplyAmbientBindingData(IBinder binder, IDictionary<string, string> bindingData)
+        {
+            // TEMP: Dig the ambient binding data out of the binder
+            FieldInfo fieldInfo = binder.GetType().GetField("_bindingSource", BindingFlags.NonPublic | BindingFlags.Instance);
+            var bindingSource = fieldInfo.GetValue(binder);
+            PropertyInfo propertyInfo = bindingSource.GetType().GetProperty("AmbientBindingContext");
+            var ambientBindingContext = propertyInfo.GetValue(bindingSource);
+            propertyInfo = ambientBindingContext.GetType().GetProperty("BindingData");
```

Comment:
I created https://github.com/Azure/azure-webjobs-sdk-script/issues/76 to track this. We'll remove this.